pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go

```diff
@@ -0,0 +1,146 @@
+	startupMonitorEnabledFn func() bool,
+	eventRecorder events.Recorder) factory.Controller {
+	fd := &startupMonitorPodConditionController{
+		operatorClient:          operatorClient,
+		targetName:              targetName,
+		podLister:               kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Lister().Pods(targetNamespace),
+		startupMonitorEnabledFn: startupMonitorEnabledFn,
+	}
+	return factory.New().WithSync(fd.sync).ResyncEvery(6*time.Minute).WithInformers(kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Informer()).ToController("StartupMonitorPodCondition", eventRecorder)
+}
+
+func (fd *startupMonitorPodConditionController) sync(_ context.Context, _ factory.SyncContext) (err error) {
+	startupPodDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodDegraded", Status: operatorv1.ConditionFalse}
+	startupPodContainerExcessiveRestartsDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodContainerExcessiveRestartsDegraded", Status: operatorv1.ConditionFalse}
+
+	defer func() {
+		if err == nil {
+			if _, _, updateError := operatorv1helpers.UpdateStatus(fd.operatorClient,
+				operatorv1helpers.UpdateConditionFn(*startupPodDegraded),
+				operatorv1helpers.UpdateConditionFn(*startupPodContainerExcessiveRestartsDegraded)); updateError != nil {
+				err = updateError
+			}
+		}
+	}()
+
+	// in practice we rely on operators to provide
+	// a condition for checking we are running on a single node cluster
+	if !fd.startupMonitorEnabledFn() {
+		return nil
+	}
+
+	monitorPod, err := fd.podLister.Get(fmt.Sprintf("%s-startup-monitor", fd.targetName))
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err // unknown error
+	}
+
+	withDegradedWhenInPendingPhaseFor(monitorPod, 2*time.Minute, startupPodDegraded)
+	withDegradedWhenHasFailedContainer(monitorPod, startupPodDegraded)
+	withDegradedWhenHasExcessiveRestarts(monitorPod, 2, startupPodContainerExcessiveRestartsDegraded)
+
+	return nil
+}
+
+func withDegradedWhenInPendingPhaseFor(monitorPod *corev1.Pod, maxPendingDuration time.Duration, startupPodDegraded *operatorv1.OperatorCondition) {
+	if monitorPod.Status.Phase != corev1.PodPending || monitorPod.Status.StartTime == nil || time.Now().Sub(monitorPod.Status.StartTime.Time) < maxPendingDuration {
+		return
+	}
+
+	startupPodDegraded.Status = operatorv1.ConditionTrue
+	pendingPodReason, pendingPodMessage := "Unknown", "unknown"
+	if len(monitorPod.Status.Reason) > 0 {
+		pendingPodReason = monitorPod.Status.Reason
+	}
+	if len(monitorPod.Status.Message) > 0 {
+		pendingPodMessage = monitorPod.Status.Message
+	}
+	startupPodDegraded.Reason = pendingPodReason
+	startupPodDegraded.Message = fmt.Sprintf("the pod %v has been in %v phase for %v due to %v, the max tolerated value is %v", monitorPod.Name, corev1.PodPending, time.Now().Sub(monitorPod.Status.StartTime.Time), pendingPodMessage, maxPendingDuration)
```

Comment:
don't use durations in messages. Messages must be constant.

Reply:
actually, the time in the message will be different than the one used in the `if`